tsl/src/nodes/vector_agg/grouping_policy_hash.c

```diff
@@ -0,0 +1,735 @@
+{
+	GroupingPolicyHash *policy = (GroupingPolicyHash *) obj;
+
+	MemoryContextReset(policy->agg_extra_mctx);
+
+	policy->returning_results = false;
+
+	h_reset(policy->table);
+	policy->have_null_key = false;
+
+	policy->stat_input_valid_rows = 0;
+	policy->stat_input_total_rows = 0;
+	policy->stat_bulk_filtered_rows = 0;
+}
+
+static void
+compute_single_aggregate(DecompressBatchState *batch_state, int start_row, int end_row,
+						 VectorAggDef *agg_def, void *agg_states, uint32 *offsets,
+						 MemoryContext agg_extra_mctx)
+{
+	ArrowArray *arg_arrow = NULL;
+	Datum arg_datum = 0;
+	bool arg_isnull = true;
+
+	/*
+	 * We have functions with one argument, and one function with no arguments
+	 * (count(*)). Collect the arguments.
+	 */
+	if (agg_def->input_offset >= 0)
+	{
+		CompressedColumnValues *values = &batch_state->compressed_columns[agg_def->input_offset];
+		Assert(values->decompression_type != DT_Invalid);
+		Assert(values->decompression_type != DT_Iterator);
+
+		if (values->arrow != NULL)
+		{
+			arg_arrow = values->arrow;
+		}
+		else
+		{
+			Assert(values->decompression_type == DT_Scalar);
+			arg_datum = *values->output_value;
+			arg_isnull = *values->output_isnull;
+		}
+	}
+
+	/*
+	 * Now call the function.
+	 */
+	if (arg_arrow != NULL)
+	{
+		/* Arrow argument. */
+		agg_def->func
+			.agg_many_vector(agg_states, offsets, start_row, end_row, arg_arrow, agg_extra_mctx);
+	}
+	else
+	{
+		/*
+		 * Scalar argument, or count(*). The latter has an optimized
+		 * implementation.
+		 */
+		if (agg_def->func.agg_many_scalar != NULL)
+		{
+			agg_def->func.agg_many_scalar(agg_states,
+										  offsets,
+										  start_row,
+										  end_row,
+										  arg_datum,
+										  arg_isnull,
+										  agg_extra_mctx);
+		}
+		else
+		{
+			for (int i = start_row; i < end_row; i++)
+			{
+				if (offsets[i] == 0)
+				{
+					continue;
+				}
+
+				void *state = (offsets[i] * agg_def->func.state_bytes + (char *) agg_states);
+				agg_def->func.agg_scalar(state, arg_datum, arg_isnull, 1, agg_extra_mctx);
+			}
+		}
+	}
+}
+
+/*
+ * Fill the aggregation state offsets for all rows using a hash table.
+ */
+static pg_attribute_always_inline uint32
+fill_offsets_impl(GroupingPolicyHash *policy, CompressedColumnValues column,
+				  const uint64 *restrict filter, uint32 next_unused_state_index, int start_row,
+				  int end_row, uint32 *restrict offsets,
+				  void (*get_key)(CompressedColumnValues column, int row, Datum *restrict key,
+								  bool *restrict valid))
+{
+	struct h_hash *restrict table = policy->table;
+	for (int row = start_row; row < end_row; row++)
+	{
+		bool key_valid = false;
+		Datum key = { 0 };
+		get_key(column, row, &key, &key_valid);
+
+		if (!arrow_row_is_valid(filter, row))
+		{
+			continue;
+		}
+
+		if (key_valid)
+		{
+			bool found = false;
+			HashEntry *restrict entry = h_insert(table, key, &found);
+			if (!found)
+			{
+				entry->agg_state_index = next_unused_state_index++;
```

Comment:
I am wondering why we're not initializing the aggregate states here instead of in an additional loop after the fill_offsets?

Reply:
Just following a common rule of thumb for high-throughput data processing: colocate similar items in memory, do the same operation on them in bulk.

In this grouping agorithm we have two major stages:
* Match input row to a unique key index (`offsets`). This uses the hash table and the key column data.
* Calculate the aggregate functions. This uses their input columns and their states.
These stages are done separately, and the data required for them are grouped together. According to the same principle, each aggregate function is processed for the entire batch, not all aggregate functions for one row.